### ISSUE TITLE
Cannot cancel an XMLHttpRequest in @cycle/http

### ISSUE THREAD

**Issue by artfuldev** [cyclejs/cyclejs#355]
With the addition of the runStreamAdapter, there is a default listener added, which takes away the ability to cancel XHRs.

**Comment by staltz** [comment]
Thanks for reporting, seems serious.

**Comment by artfuldev** [comment]
@staltz was not able to submit my entry to your flux-challenge because of this. Was breaking my head over why until @Hypnosphi found this.

**Comment by Hypnosphi** [comment]
The problem is that there are some cases when we want requests to complete even if we don't listen to responses (some POST requests, for example). Maybe `makeHTTPDriver` should accept some `{lazy: true}` option which will indicate that we're going to listen to all the `response$` explicitly.

**Comment by staltz** [comment]
I have a proposal for this.

```js
  let request$ = xs.of({
    url: 'http://localhost:8080/hello', // GET method by default
    category: 'hello',
    cancelWhen: xs.periodic(2000).take(1),
  });
```

Cancels the XHR after 2 seconds.

This would be a non-breaking change, which is good. What do people think?
Also @TylorS

**Comment by eshansingh** [comment]
So cancelWhen takes a stream, that, when completed, cancels the XHR. Am I getting this?

**Comment by staltz** [comment]
Takes a stream that when it emits `next` *or* `complete`, cancels the XHR.

**Comment by eshansingh** [comment]
It is a really good idea, and it's even better that its non-breaking! :+1: from me.

**Comment by Hypnosphi** [comment]
I don't think I can really know in advance what will be the reason of cancellation. Let's look at cycle HTTP docs:

```js
let httpResponse$ = httpResponse$$.flatten(); // flattens the metastream
  // the reason why we need to flatten in this API is that you
  // should choose which concurrency strategy to use.
  // Normal xstream flatten() has limited concurrency of 1, meaning that
  // the previous request will be canceled once the next request to the
  // same resource occurs.
  // To have full concurrency (no cancelling), use flattenConcurrently()
```

It's natural to expect here, that the bare fact that we stop listening to previous request leads to its cancellation. In fact this functionality seems to be [provided](https://github.com/cyclejs/cyclejs/blob/master/http/src/http-driver.ts#L100), but it's not in real, because of the driver's listener presence.
So how do you achieve the same functionality with `cancelWhen` field on the request object?

**Comment by staltz** [comment]
Yeah, you're right, the driver's noop listener is a problem. So far we know these requirements for sure:

- POST request without a response$ listener should execute no matter what
- If the above has a `cancelWhen` field, then it should execute until the cancellation stream emits

Here's another addition to the proposal:

```diff
  let request$ = xs.of({
    url: 'http://localhost:8080/hello',
    method: 'POST',
    category: 'hello',
-   cancelWhen: xs.periodic(2000).take(1),
+   cancelWhen: () => xs.periodic(2000).take(1),
  });
```

Which would enable:
```js
  let request$ = xs.of({
    url: 'http://localhost:8080/hello',
    method: 'POST',
    category: 'hello',
    // cancel when the next request happens
    // (without drop(1) it would cancel as soon as THIS request happens)
    cancelWhen: () => request$.drop(1),
  });
```

**Comment by staltz** [comment]
To debate with myself, here's a problem with `cancelWhen` proposal: won't work well in the DevTool visualizer, because the cancellation stream isn't hooked to a specific sink, it's just embedded in the request object.

**Comment by Hypnosphi** [comment]
So if later I want to switch to
```js
let httpResponse$ = HTTP.select('hello').flatten()
```

I have to replicate the filtering inside `cancelWhen`:

```js
cancelWhen: () => request$.filter(request => request.category === 'hello').drop(1)
```

**Comment by Hypnosphi** [comment]
And this `drop(1)` is needed if `request$` is a MemoryStream, and not needed otherwise. Too much things to consider comparing to just one `makeHTTPDriver` option

**Comment by artfuldev** [comment]
@staltz Yes, it's better to have a separate stream that sinks into the httpDriver that cancels requests, in that case.
Here's a proposal:
HTTPSource  constructor(sink$: Stream<string | HTTPRequest | HTTPCancellationRequest>)

**Comment by artfuldev** [comment]
We can write code such that HTTPCancellationRequest removes the noop listener. The HTTPCancellationRequest will contain some kind of `filter` object which can filter the responses that it should cancel..

**Comment by Hypnosphi** [comment]
A hybrid solution would be `{lazy: true}` on request object and
```
if (!reqInput.lazy) {
  response$.addListener(...);
}
```

**Comment by staltz** [comment]
Yeah @Hypnosphi, that probably would work and still be conservative enough.

I'm still giving some time and thought to this issue though.

**Comment by eshansingh** [comment]
So, what's our final take on this?

**Comment by bumi001** [comment]
How about:
  const evE = new EventEmitter();
  let request$ = xs.of({
    url: 'http://localhost:8080/hello',
    method: 'POST',
    category: 'hello',
    cancelWhen: fromEvent(evE, 'cancel'),  // <---- add a listener inside 'of' if cancelWhen is of type
                                                             //EventEmitter
  });

//somewhere else in code
evE.emit('cancel', 2); //cancel in 2 seconds

**Comment by bumi001** [comment]
Please ignore my comment. cancelWhen would be a stream!

**Comment by whitecolor** [comment]
1) solution to would introduce `lazy` param and make lazy requests obviously cancellable - so request starts on  subscription and requests is canceled if subscription ends before request completion - it is definitely logically correct.

2) for eager requests start performing effect without any subscription, and so there would be no need to create internal subscription to `response$`, and cancel it if subscription appears and ends - but **I'm not sure if it is logically correct**, that request starts regardless subscription presence and canceled regarding subscription end.

**Comment by staltz** [comment]
Released this in version 10.1.0 thanks to @Hypnosphi
